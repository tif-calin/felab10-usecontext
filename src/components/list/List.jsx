import React from 'react';
/* import PropTypes from 'prop-types'; */
import { useLocation } from 'react-router-dom';
import Header from '../header/Header';
import { useCharacters } from '../../state/listsOfThings.js';
import APISwitcher from './APISwitcher';

const List = () => {

  const api = new URLSearchParams(useLocation().search).get('api');
  const state = useCharacters(api); 
  const { loading, characters, API, page, setPage, setAPI } = state;

  return <>
    <Header/>
    <APISwitcher setAPI={setAPI}/>
    <hr/>
    <h3>{API.name}</h3>
    {loading
      ? <div>Loading...</div>
      : <>
        <form>
          <button 
            onClick={() => setPage(Math.max(page - 1, 1))}
          >&#x2190;</button>
          <span>{page}</span>
          <button 
            onClick={() => setPage(Math.min(page + 1, ~~(API.count / 20)))}
          >&#x2192;</button>
        </form>

        <ul className="List">
          {characters.map(character => <li key={character[API.id]}>
            <img src={character[API.image]} alt={character[API.title]}/>
            <span>{Array.isArray(API.title) 
              ? API.title.reduce((acc, val) => acc[val], character)
              : character[API.title]
            }</span>
          </li>)}
        </ul>
      </>
    }
  </>;
};

List.propTypes = {};

export default List;
