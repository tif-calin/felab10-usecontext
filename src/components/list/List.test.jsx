import React from 'react';
import { render, cleanup, act } from '@testing-library/react';
import List from './List';
import { MemoryRouter } from 'react-router-dom';

describe('List component', () => {

  afterEach(() => cleanup());

  test('snpashot test for List', () => {
    const { asFragment } = render(<MemoryRouter>
      <List match={{ params: { api: 'avatar' } }}/>
    </MemoryRouter>);

    expect(asFragment()).toMatchSnapshot();
  });

  test('user clicks on the next page', async () => {
    const { findByText } = await render(<MemoryRouter>
      <List match={{ params: { api: 'avatar' } }}/>
    </MemoryRouter>);

    const nextPageButton = await findByText('â†’');

    act(() => nextPageButton.click());

    expect(await findByText('Banished servant')).toBeInTheDocument();
  });
  
});
