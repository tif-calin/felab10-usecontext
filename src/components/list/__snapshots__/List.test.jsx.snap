// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component snpashot test for List 1`] = `
<DocumentFragment>
  <header
    class="Header"
  >
    <h1>
      Lists of Things
    </h1>
    <nav>
      <a
        href="/"
      >
        home
      </a>
    </nav>
  </header>
  <div>
    Loading...
  </div>
</DocumentFragment>
`;
